File: sdk/typescript/src/server/assets.ts
import { listQuery, type HttpTransport } from "../http.js";
import type {
  AssetDef,
  AssetHolding,
  AssetLedgerEntry,
  AssetOpResponse,
  ListParams,
  ReconcileResponse,
} from "../types.js";

export class ServerAssetsService {
  constructor(private readonly http: HttpTransport) {}

  async createAssetDef(input: {
    code: string;
    name: string;
    class: string;
    decimals?: number;
    max_quantity?: string;
    expires_in?: string;
    tradable?: boolean;
    unique_per_owner?: boolean;
    upgradeable?: boolean;
    metadata?: Record<string, unknown>;
  }): Promise<AssetDef> {
    return this.http.request<AssetDef>("POST", "/v1/server/assets/defs", {
      auth: "apiKey",
      body: input,
    });
  }

  async listAssetDefs(params?: ListParams): Promise<AssetDef[]> {
    const res = await this.http.request<{ defs: AssetDef[] }>("GET", "/v1/server/assets/defs", {
      auth: "apiKey",
      query: listQuery(params),
    });
    return res.defs ?? [];
  }

  async getAssetDef(defId: string): Promise<AssetDef> {
    return this.http.request<AssetDef>("GET", `/v1/server/assets/defs/${defId}`, { auth: "apiKey" });
  }

  async updateAssetDef(
    defId: string,
    input: {
      name?: string;
      decimals?: number;
      max_quantity?: string;
      expires_in?: string;
      tradable?: boolean;
      unique_per_owner?: boolean;
      upgradeable?: boolean;
      metadata?: Record<string, unknown>;
      status?: string;
    }
  ): Promise<AssetDef> {
    return this.http.request<AssetDef>("PATCH", `/v1/server/assets/defs/${defId}`, {
      auth: "apiKey",
      body: { def_id: defId, ...input },
    });
  }

  async deleteAssetDef(defId: string): Promise<void> {
    await this.http.request<void>("DELETE", `/v1/server/assets/defs/${defId}`, { auth: "apiKey" });
  }

  async grant(input: {
    owner_id: string;
    def_code: string;
    quantity: string;
    idempotency_key: string;
    expires_at?: string;
    level?: number;
    metadata?: Record<string, unknown>;
    ref_type?: string;
    ref_id?: string;
  }): Promise<AssetOpResponse> {
    return this.http.request<AssetOpResponse>("POST", "/v1/server/assets:grant", {
      auth: "apiKey",
      body: input,
    });
  }

  async consume(input: {
    owner_id: string;
    def_code: string;
    quantity: string;
    idempotency_key: string;
    ref_type?: string;
    ref_id?: string;
  }): Promise<AssetOpResponse> {
    return this.http.request<AssetOpResponse>("POST", "/v1/server/assets:consume", {
      auth: "apiKey",
      body: input,
    });
  }

  async transfer(input: {
    from_owner_id: string;
    to_owner_id: string;
    def_code: string;
    quantity: string;
    idempotency_key: string;
    ref_type?: string;
    ref_id?: string;
  }): Promise<AssetOpResponse> {
    return this.http.request<AssetOpResponse>("POST", "/v1/server/assets:transfer", {
      auth: "apiKey",
      body: input,
    });
  }

  async mutate(input: {
    holding_id: string;
    idempotency_key: string;
    level?: number;
    expires_at?: string;
    metadata?: Record<string, unknown>;
    ref_type?: string;
    ref_id?: string;
  }): Promise<AssetOpResponse> {
    return this.http.request<AssetOpResponse>("POST", "/v1/server/assets:mutate", {
      auth: "apiKey",
      body: input,
    });
  }

  async expire(input: { holding_id: string; idempotency_key: string }): Promise<AssetOpResponse> {
    return this.http.request<AssetOpResponse>("POST", "/v1/server/assets:expire", {
      auth: "apiKey",
      body: input,
    });
  }

  async reconcile(): Promise<ReconcileResponse> {
    return this.http.request<ReconcileResponse>("POST", "/v1/server/assets:reconcile", {
      auth: "apiKey",
      body: {},
    });
  }

  async listUserAssets(ownerId: string, params?: ListParams): Promise<AssetHolding[]> {
    const res = await this.http.request<{ holdings: AssetHolding[] }>(
      "GET",
      `/v1/server/assets/users/${ownerId}`,
      { auth: "apiKey", query: listQuery(params) }
    );
    return res.holdings ?? [];
  }

  async listUserLedger(
    ownerId: string,
    params?: ListParams & { def_code?: string }
  ): Promise<AssetLedgerEntry[]> {
    const res = await this.http.request<{ entries: AssetLedgerEntry[] }>(
      "GET",
      `/v1/server/assets/users/${ownerId}/ledger`,
      {
        auth: "apiKey",
        query: { ...listQuery(params), def_code: params?.def_code },
      }
    );
    return res.entries ?? [];
  }
}

File: sdk/go/server/assets.go
package server

import (
	"context"

	serverv1 "github.com/torchwooddev/torchwood/genproto/server/v1"
	sharedv1 "github.com/torchwooddev/torchwood/genproto/shared/v1"
)

// AssetsService 封装 Server API 的资产管理（Def CRUD + 五动词 + 用户资产查询）。
// 写方法仅 system / admin-key；SDK 不提供终端用户写入口。
type AssetsService struct {
	c   *Client
	api serverv1.AssetsServiceClient
}

func (s *AssetsService) CreateAssetDef(ctx context.Context, req *serverv1.CreateAssetDefRequest) (*serverv1.AssetDef, error) {
	return s.api.CreateAssetDef(ctx, req)
}

func (s *AssetsService) ListAssetDefs(ctx context.Context, req *sharedv1.ListRequest) (*serverv1.ListAssetDefsResponse, error) {
	return s.api.ListAssetDefs(ctx, req)
}

func (s *AssetsService) GetAssetDef(ctx context.Context, defID string) (*serverv1.AssetDef, error) {
	return s.api.GetAssetDef(ctx, &serverv1.GetAssetDefRequest{DefId: defID})
}

func (s *AssetsService) UpdateAssetDef(ctx context.Context, req *serverv1.UpdateAssetDefRequest) (*serverv1.AssetDef, error) {
	return s.api.UpdateAssetDef(ctx, req)
}

func (s *AssetsService) DeleteAssetDef(ctx context.Context, defID string) error {
	_, err := s.api.DeleteAssetDef(ctx, &serverv1.DeleteAssetDefRequest{DefId: defID})
	return err
}

func (s *AssetsService) Grant(ctx context.Context, req *serverv1.GrantRequest) (*serverv1.AssetOpResponse, error) {
	return s.api.Grant(ctx, req)
}

func (s *AssetsService) Consume(ctx context.Context, req *serverv1.ConsumeRequest) (*serverv1.AssetOpResponse, error) {
	return s.api.Consume(ctx, req)
}

func (s *AssetsService) Transfer(ctx context.Context, req *serverv1.TransferRequest) (*serverv1.AssetOpResponse, error) {
	return s.api.Transfer(ctx, req)
}

func (s *AssetsService) Mutate(ctx context.Context, req *serverv1.MutateRequest) (*serverv1.AssetOpResponse, error) {
	return s.api.Mutate(ctx, req)
}

func (s *AssetsService) Expire(ctx context.Context, req *serverv1.ExpireRequest) (*serverv1.AssetOpResponse, error) {
	return s.api.Expire(ctx, req)
}

func (s *AssetsService) Reconcile(ctx context.Context) (*serverv1.ReconcileResponse, error) {
	return s.api.Reconcile(ctx, &serverv1.ReconcileRequest{})
}

func (s *AssetsService) ListUserAssets(ctx context.Context, req *serverv1.ListUserAssetsRequest) (*serverv1.ListUserAssetsResponse, error) {
	return s.api.ListUserAssets(ctx, req)
}

func (s *AssetsService) ListUserLedger(ctx context.Context, req *serverv1.ListUserLedgerRequest) (*serverv1.ListUserLedgerResponse, error) {
	return s.api.ListUserLedger(ctx, req)
}
